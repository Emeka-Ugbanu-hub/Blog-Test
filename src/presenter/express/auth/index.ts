import login from './login';
import register from './register';
import forgetPassword from './forgetPassword';

export {
  login, 
  register,
  forgetPassword
}