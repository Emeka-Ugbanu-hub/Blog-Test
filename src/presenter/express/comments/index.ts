import createComment from './createComment';
import getCommentById from './getCommentById';
export {
  createComment,
  getCommentById
}