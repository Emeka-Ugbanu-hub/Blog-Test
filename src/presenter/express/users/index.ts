import getUserById from './getUserById';
import getUsers from './getUsers';
import updateUser from './updateUser';
export {
  getUserById,
  getUsers,
  updateUser
}