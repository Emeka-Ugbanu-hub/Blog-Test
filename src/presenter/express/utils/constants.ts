const OK_200_HTTP_CODE = 200;
const CREATED_201_HTTP_CODE = 201;
const NO_CONTENT_204_HTTP_CODE = 204;
const BAD_REQUEST_400_HTTP_CODE = 400;
const UNAUTHORISED_401_HTTP_CODE = 401;
const FORBIDDEN_403_HTTP_CODE = 403;
const NOT_FOUND_404_HTTP_CODE = 404;
const CONFLICT_409_HTTP_CODE = 409;
const UNPROCESSABLE_ENTITY_422_HTTP_CODE = 422;
const SERVER_ERROR_500_HTTP_CODE = 500;


export {
  OK_200_HTTP_CODE,
  CREATED_201_HTTP_CODE,
  NO_CONTENT_204_HTTP_CODE,
  BAD_REQUEST_400_HTTP_CODE,
  UNAUTHORISED_401_HTTP_CODE,
  FORBIDDEN_403_HTTP_CODE,
  NOT_FOUND_404_HTTP_CODE,
  CONFLICT_409_HTTP_CODE,
  UNPROCESSABLE_ENTITY_422_HTTP_CODE,
  SERVER_ERROR_500_HTTP_CODE,
}