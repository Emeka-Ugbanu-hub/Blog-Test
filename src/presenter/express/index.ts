import { Router, Request, Response} from 'express';
import Config from './Config';
import enhancedRouter from './enhancedRouter';
import {register, login, forgetPassword,resetPassword} from './auth';
import {getPosts, createPost, getPostById, updatePost, deletePostById} from './posts';
import {getUserById, getUsers, updateUser, createUser, deleteUserById} from './users';
import {createComment, getCommentById, getComments, updateComment} from './comments';

export default (config: Config): Router => {
  const router: Router = enhancedRouter(config);

  router.get('/', (req: Request, res: Response) => {
    res.status(200)
       .json({message: "This is where the awesomeness happen..."});
  });

  router.post('/auth/register', register(config));
  router.post('/auth/login', login(config));
  router.post('/auth/forget-password', forgetPassword(config));
  router.post('/auth/reset-password', resetPassword(config));

  router.post('/users', createUser(config));
  router.delete('/users/:user_id', deleteUserById(config));
  router.get('/users/:user_id', getUserById(config));
  router.get('/users', getUsers(config));
  router.patch('/users/:user_id', updateUser(config));
  
  router.get('/posts', getPosts(config));
  router.get('/posts/:post_id', getPostById(config));
  router.delete('/posts/:post_id', deletePostById(config));
  router.post('/posts', createPost(config));
  router.patch('/posts/:post_id', updatePost(config));

  router.get('/comments', getComments(config));
  router.post('/comments', createComment(config));
  router.get('/comments/:comment_id', getCommentById(config));
  router.patch('/comments/:comment_id', updateComment(config));

  return router;
}