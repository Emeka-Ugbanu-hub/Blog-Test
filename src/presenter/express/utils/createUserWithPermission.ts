import Service from '../../../service/Service';
import {TEST_VALID_REGIRSTER_USER,TEST_DEFAULT_ROLE_NAME } from '../../../utils/testValues';


export default (service: Service, permissionName: string): Promise<any> => {
  return new Promise(async(resolve,reject)=>{
    const user: any = await service.createUser(TEST_VALID_REGIRSTER_USER);
    const role = { name: TEST_DEFAULT_ROLE_NAME };
    const permission = { name: permissionName };
    await service.createUserPermissions({userId: user.id, permissions: [permission], role });
    resolve(user);
  });
 
}