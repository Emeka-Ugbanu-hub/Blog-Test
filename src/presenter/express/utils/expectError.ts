import {UNPROCESSABLE_ENTITY_422_HTTP_CODE} from './constants';

interface Response {
  status: number;
  body: any;
}

export default (response: Response, expectedStatusCode: number = UNPROCESSABLE_ENTITY_422_HTTP_CODE) => {
  expect(response.status).toBe(expectedStatusCode);
  expect(response.body).toMatchSnapshot();
}