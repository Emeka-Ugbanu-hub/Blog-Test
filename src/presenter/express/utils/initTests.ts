import config from '../../../config';
import * as sourceMapSupport from 'source-map-support';
sourceMapSupport.install();
import * as iconvLite from 'iconv-lite';
iconvLite.encodingExists('foo');
import * as express from 'express';
import * as createSupertest from 'supertest';
import logger from '../../../logger';
import presenter from '../index';
import serviceFactory from '../../../service/factory';
import {API_ROUTE_V1} from '../../../utils/constants';

const app: express.Application = express();

const service = serviceFactory();

const presenterFacade = presenter({
  morganLogFormat: config.express.morganLogFormat,
  morganDirectory: config.express.morganDirectory,
  service,
  logger
});

app.use(API_ROUTE_V1, presenterFacade);

const request = createSupertest(app);

export default () => {

  beforeAll(async() => {
    await service.migrate();
  });

  beforeEach(async() => {
    await service.migrate();
  });
  
  afterEach(async() => {
    await service.rollback();
  });

  afterAll(async() => {
    await service.clearService();
  });
  
  return { service, request };
};