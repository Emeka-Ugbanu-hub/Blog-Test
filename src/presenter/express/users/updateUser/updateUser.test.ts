import initTests from '../../utils/initTests';
import {API_ROUTE_V1} from '../../../../utils/constants';
import {Response} from 'express';
import { OK, FORBIDDEN, UNAUTHORIZED, NOT_FOUND, UNPROCESSABLE_ENTITY } from 'http-status-codes';
import config from '../../../../config';
import {TEST_INVALID_JWT_TOKEN, TEST_INVALID_EMAIL,TEST_DIFFERENT_VALID_EMAIL, TEST_DIFFERENT_VALID_PASSWORD,TEST_VALID_PASSWORD, TEST_TOO_SHORT_PASSWORD,TEST_VALID_ANOTHER_REGIRSTER_USER, TEST_VALID_REGISTER_USER } from '../../../../utils/testValues';
import expectError from '../../utils/expectError';
import generateJwtToken from '../../../../utils/jwt/generateToken';
import createUserWithPermission from '../../utils/createUserWithPermission';
import {CAN_UPDATE_USER, DEFAULT_USERS_PAGINATION_LIMIT, DEFAULT_USERS_PAGINATION_OFFSET} from '../../../../utils/constants';
import {fakeUsers} from '../../../../utils/fakesFactory';
import verifyPassword from '../../../../utils/verifyPassword';
import * as moment from 'moment';

describe(__filename, () => {

  const { service, request } = initTests();

  it('should fail to update user when unauthenticated', async () => {
    const response = await request.patch(`${API_ROUTE_V1}/users/1`);
    expectError(response, UNAUTHORIZED);
  });

  it('should fail to update user when invalid token provided in authorization header', async () => {
    const response = await request.patch(`${API_ROUTE_V1}/users/1`)
                                  .set('Authorization' , TEST_INVALID_JWT_TOKEN);
    expectError(response, UNAUTHORIZED);
  });

  it('should fail to update user when insufficent permissions', async () => {
    const userWithoutPermissions = await service.createUser(TEST_VALID_REGISTER_USER);
    const validToken = await generateJwtToken({data: {id: userWithoutPermissions.id}});
    const response = await request.patch(`${API_ROUTE_V1}/users/${userWithoutPermissions.id}`)
                                  .set('Authorization' , validToken);
    expectError(response, FORBIDDEN);
  });

  it('should fail to update user with invalid user email', async () => {
    const user = await createUserWithPermission(service, CAN_UPDATE_USER);
    
    const validToken = await generateJwtToken({data: {id: user.id}});
    const response = await request.patch(`${API_ROUTE_V1}/users/${user.id}`)
                                  .set('Authorization' , validToken)
                                  .send({
                                    email: TEST_INVALID_EMAIL
                                  });

    expectError(response, UNPROCESSABLE_ENTITY);
  });

  it('should fail to update user with too short password', async () => {
    const user = await createUserWithPermission(service, CAN_UPDATE_USER);
    
    const validToken = await generateJwtToken({data: {id: user.id}});
    const response = await request.patch(`${API_ROUTE_V1}/users/${user.id}`)
                                  .set('Authorization' , validToken)
                                  .send({
                                    password: TEST_TOO_SHORT_PASSWORD
                                  });

    expectError(response, UNPROCESSABLE_ENTITY);
  });


  it('should fail to update user without password confirmation', async () => {
    const user = await createUserWithPermission(service, CAN_UPDATE_USER);
    
    const validToken = await generateJwtToken({data: {id: user.id}});
    const response = await request.patch(`${API_ROUTE_V1}/users/${user.id}`)
                                  .set('Authorization' , validToken)
                                  .send({
                                    password: TEST_VALID_PASSWORD
                                  });
    expectError(response, UNPROCESSABLE_ENTITY);
  });

  it('should fail to update user when password and password_confirmation don\'t match', async () => {
    const user = await createUserWithPermission(service, CAN_UPDATE_USER);
    
    const validToken = await generateJwtToken({data: {id: user.id}});
    const response = await request.patch(`${API_ROUTE_V1}/users/${user.id}`)
                                  .set('Authorization' , validToken)
                                  .send({
                                    password: TEST_VALID_PASSWORD,
                                    password_confirmation: TEST_DIFFERENT_VALID_PASSWORD
                                  });
    expectError(response, UNPROCESSABLE_ENTITY);
  });

  it('should fail to update user when password and password_confirmation don\'t match', async () => {
    const user = await createUserWithPermission(service, CAN_UPDATE_USER);
    
    const validToken = await generateJwtToken({data: {id: user.id}});
    const response = await request.patch(`${API_ROUTE_V1}/users/${user.id}`)
                                  .set('Authorization' , validToken)
                                  .send({
                                    password: TEST_VALID_PASSWORD,
                                    password_confirmation: TEST_DIFFERENT_VALID_PASSWORD
                                  });
    expectError(response, UNPROCESSABLE_ENTITY);
  });

  it('should successfuly update user when password and password_confirmation match', async () => {
    const user = await createUserWithPermission(service, CAN_UPDATE_USER);
    const newUser = fakeUsers({count: 1, only: [
      'email', 'firstname', 'password', 'lastname', 'bio', 'password_confirmation'
    ]});
    const validToken = await generateJwtToken({data: {id: user.id}});
    const response = await request.patch(`${API_ROUTE_V1}/users/${user.id}`)
                                  .set('Authorization' , validToken)
                                  .send(newUser);

    const updatedUser: any = await service.getUserById({id: user.id});  
    const match = await verifyPassword(newUser.password, updatedUser.password); 

    expect(match).toBe(true);
    const now = moment(new Date());
    const correctUpdateAt = moment.duration(now.diff(response.body.updated_at)).asMilliseconds() < 10000;
    expect(response.body.email).toEqual(newUser.email);   
    expect(response.body.bio).toEqual(newUser.bio);   
    expect(response.body.firstname).toEqual(newUser.firstname);   
    expect(response.body.lastname).toEqual(newUser.lastname);         
    expect(correctUpdateAt).toBe(true);
    expect(response.status).toBe(OK);
  });
});