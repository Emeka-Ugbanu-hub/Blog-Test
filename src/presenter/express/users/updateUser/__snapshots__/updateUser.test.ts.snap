// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/users/updateUser/updateUser.test.ts should fail to get user when insufficent permissions 1`] = `
Object {
  "message": "Forbidden",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/users/updateUser/updateUser.test.ts should fail to get user when unauthenticated 1`] = `
Object {
  "message": "Missing JWT token",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/users/updateUser/updateUser.test.ts should fail to get users when invalid token provided in authorization header 1`] = `
Object {
  "message": "Missing JWT token",
}
`;
